### ISSUE TITLE
Retrieve layer/model weights

### ISSUE THREAD

**Issue by tanmayb123** [tensorflow/swift-apis#25]
There should be an easy way to extract the weights of a Layer/Model and save them. Currently, I've got the not-so-pretty method of doing this:

```
public func weights() -> [Tensor<Float>] {
    return [dense1.weight, dense1.bias, dense2.weight, dense2.bias, dense3.weight, dense3.bias, dense4.weight, dense4.bias]
}
```

Originally, I thought of doing something like adding a `weights() -> [Tensor<Scalar>]` function to the `Layer` protocol. For example, `Dense<Scalar: TensorFlowFloatingPoint>` would implement it like so:

```
public func weights() -> [Tensor<Scalar>] {
    return [weight, bias]
}
```

My layer could then do this:

```
public func weights() -> [Tensor<Scalar>] {
    return dense1.weights() +
           dense2.weights() +
           dense3.weights() +
           dense4.weights()
}
```

Is there an easier way to do this?

**Comment by rxwei** [comment]
Forming an arrays of tensors is never a good solution -- it performs unnecessary semantic copies and keeps more tensors alive. Our solution is key paths.  Layers conform to `KeyPathIterable`. You can take a look at `Optimizers.swift` to see how they iterate over all weights using key paths. 

```swift
for kp in model.allDifferentiableVariables.recursivelyAllKeyPath(to: Tensor<Float>.self) {
    write(model.allDifferentiableVariables[kp], to: ...)
}
```

We are going to provide a general serialization/deserialization API that works well with the [`Codable`](https://developer.apple.com/documentation/foundation/archives_and_serialization/encoding_and_decoding_custom_types) mechanism and that can support model saving and checkpointing. One can implement `Codable` conformances using key paths. Feel free to experiment with it if you'd like. We plan to look into this in a few weeks (after the dev summit).

**Comment by Shashi456** [comment]
@rxwei I'm extremely keen to work on this, with #117 and #76 we will be done with pooling layer implementations. We have almost all layers needed for most basic classification networks. Although before implementation, are there things you would recommend to read up on and then start a discussion here to decide on some sort of specification for the implementation?

**Comment by thommackey** [comment]
Exciting that this is being discussed! It's probably a little early for FRs, but I'd like to encourage the implementation to support eventual conversion/export to other formats e.g. ONNX & CoreML. I'm assuming the simplest (?) way would be to support saving a .pb file, which is one of TensorFlow's standard saved model formats. That way users could use all the existing .pb conversion tools. But it might not be the best way, I get the impression TF's saved model formats are a little overengineered.

**Comment by Oktai15** [comment]
@rxwei any update about general serialization/deserialization API? It is the main feature for any framework :)

**Comment by Shashi456** [comment]
Can we restart this discussion? Or atleast offer hack-y ways of making this possible until we work on this fulltime?

**Comment by tanmayb123** [comment]
Temporarily, I've just been iterating through `recursivelyAllWritableKeyPaths`, converting each one to a numpy array, and saving it as a large `npy` file. Then, when loading, I zip together the keypaths and items in the numpy array, and set the weights accordingly.

**Comment by rahulbhalley** [comment]
Any updates on making checkpointing/loading possible with `Codable`? Or is there any timeline decided?

**Comment by kongzii** [comment]
I created example using recursivelyAllWritableKeyPaths, for those who are interested: https://gist.github.com/kongzii/62b9d978a6536bb97095ed3fb74e30fd

**Comment by rahulbhalley** [comment]
@kongzii your approach looks hood to me (maybe not something S4TF team is aiming for). My question is can you extract and save parameters layer-wise? That could help in loading pre-trained models with recursivelyAllWritableKeyPaths.  

Thanks for your code btw.

**Comment by brettkoonce** [comment]
If you're interesting in loading/saving weights using numpy, you should check out Meir Rosendorff's blog post on the subject: https://medium.com/dvt-engineering/will-swift-take-over-python-as-the-go-to-language-for-ml-ff9f8d9d7141

**Comment by kongzii** [comment]
@rahulbhalley Thanks! All layers should conform to `Layer`, so you can execute `saveWeights` and `loadWeights` manually per layer, eg. `model.conv1d.saveWeights(...)`, if this is what you meant.

**Comment by eaplatanios** [comment]
You can also check [how that's done in the BERT PR](https://github.com/tensorflow/swift-models/pull/231/files#diff-e00a3eb6164050e7694bebc15bb229ecR781) using a TF checkpoint reader helper I added a while back.

**Comment by rahulbhalley** [comment]
Thanks @kongzii it works very nicely. Sorry, I should've thought a little more clearly that `recursivelyAllWritableKeyPaths` is a method on `Layer` protocol. 

Thank you!

**Comment by rahulbhalley** [comment]
@eaplatanios I looked at your provided [link](https://github.com/tensorflow/swift-models/pull/231/files#diff-e00a3eb6164050e7694bebc15bb229ecR781) which uses the `TensorFlowCheckpointReader`. I'm sorry I couldn't find the implementation of `TensorFlowCheckpointReader` in your [forked repository](https://github.com/eaplatanios/swift-models/tree/bert). The closest I found was the [official implementation](https://github.com/tensorflow/swift-models/blob/dfa2c88d951a4f9ba720284ac1b327a60d569422/MiniGo/Models/PythonCheckpointReader.swift) but it does not have any `loadTensor` method you used. All it has is layer specific `load` method which further calls `PythonCheckpointReader`'s `readTensor(layerName: weightName)` method. 

Please provide a direct link to your implementation code for `TensorFlowCheckpointReader`.

**Comment by eaplatanios** [comment]
@rahulbhalley It's been pushed [here](https://github.com/tensorflow/swift-apis/blob/master/Sources/TensorFlow/Core/Serialization.swift). :)

**Comment by rahulbhalley** [comment]
Why isn't S4TF team trying to use ProtoBuf? I think PyTorch, ONNX, etc. uses that and even  has its own implementation [here](https://github.com/apple/swift-protobuf) in Swift itself. 

cc: @rxwei @dan-zheng

**Comment by eaplatanios** [comment]
@rahulbhalley TF checkpoints are in Protobuf format and the C API checkpoint reader is a wrapper around the C++ protobuf library. I believe @BradLarson is currently working on a pure Swift implementation using the Swift protobuf library.

**Comment by rahulbhalley** [comment]
@eaplatanios that will be cool if we get checkpointing implemented in Swift protobuf. All the best @BradLarson!

**Comment by jbmaxwell** [comment]
Any progress on this?

**Comment by BradLarson** [comment]
Apologies for not updating this thread directly, but for those who hadn't seen it, I presented recent work on loading tensor checkpoints, near-term plans for saving checkpoints, and open design questions about this at [a recent public design meeting](https://groups.google.com/a/tensorflow.org/forum/m/#!topic/swift/SYf7UJjLJqo) (slides available [here](https://docs.google.com/presentation/d/1hC4LfND_INqU1X2LE9xaVgnTYpdkopmcX7_lClTAYq0/edit?usp=sharing)).

The current status is that we now have [a native Swift implementation for TensorFlow v2 checkpoint loading](https://github.com/tensorflow/swift-models/tree/master/Support/Checkpoints) staged in swift-models and incorporated with the models there. A TensorFlow v2 checkpoint writer is under development, as is loading from SavedModel files. Beyond that, there are broader design questions for graph serialization and other aspects of this. This work is ongoing in the swift-models repository as a testbed for these designs, so you can follow along with the pull requests and commentary there.

**Comment by vkuznet** [comment]
+1

**Comment by xanderdunn** [comment]
@BradLarson thank you very much for the `CheckpointReader` and `CheckpointWriter` implementations in swift-models. I am successfully saving and loading checkpoints on my Transformer model using them.

Saving to a checkpoint looks like this:
```swift
// Based on https://github.com/tensorflow/swift-models/blob/542a03c09aa7bd6e326a18b5d086f2bec8b4b24c/Models/Text/GPT2/CheckpointWriter.swift#L68
// Removed the renaming functionality and simply keeping the paths of the variables as they are in Swift
public func recursivelyObtainTensors(_ obj: Any,
                                     scope: String? = nil,
                                     tensors: inout [String: Tensor<Float>],
                                     separator: String) {
    let m = Mirror(reflecting: obj)

    for (i, child) in m.children.enumerated() {
        let index: String = {
            if let _ = obj as? [Any] {
                return "\(i)"
            } else {
                return ""
            }
        }()
        let label = child.label != nil ? child.label! : "h" + index
        let path = (scope != nil ? scope! + separator : "") + label
        if let tensor = child.value as? Tensor<Float> {
            tensors[path] = tensor
        }
        recursivelyObtainTensors(child.value, scope: path, tensors: &tensors, separator: separator)
    }
}

// From https://github.com/tensorflow/swift-models/blob/542a03c09aa7bd6e326a18b5d086f2bec8b4b24c/Models/Text/GPT2/CheckpointWriter.swift#L68
public func saveCheckpoint<Model: Layer>(model: Model,
                                         outputPath: URL = URL(fileURLWithPath: "./data/checkpoints/"),
                                         name: String = "checkpoint") {
    var tensors = [String: Tensor<Float>]()
    recursivelyObtainTensors(model, scope: "model", tensors: &tensors, separator: "/")

    let writer = CheckpointWriter(tensors: tensors)
    do {
        try writer.write(to: outputPath, name: name)
    } catch {
        print("Failed to write checkpoint to \(outputPath): \(error)")
    }
}
```

Loading from a checkpoint looks like this:
```swift
/*
A layer should implmeent this to restore from a checkpoint reader
*/
protocol InitializableFromCheckpoint {
    init(reader: CheckpointReader, config: TransformerConfig, scope: String)
}

extension MyLinearlyProjectedInputToTransformer: InitializableFromCheckpoint {
    public init(reader: CheckpointReader, config: TransformerConfig, scope: String) {
        self.hiddenSize = config.hiddenSize
        self.sequenceLength = config.sequenceLength
        self.numVariables = config.numVariables
        self.batchSize = config.batchSize
        self.device = config.device
        self.linearProjection = Dense<Float>(weight: reader.readTensor(name: scope + "/linearProjection/weight"),
                                             bias: reader.readTensor(name: scope + "/linearProjection/bias"),
                                             activation: identity)
        self.positionEmbedding = Embedding<Float>(embeddings:
                                           reader.readTensor(name: scope + "/positionEmbedding/embeddings"))
        self.embeddingLayerNorm = LayerNorm<Float>(offset:
                                                   reader.readTensor(name: scope + "/embeddingLayerNorm/offset"),
                                                   scale:
                                                   reader.readTensor(name: scope + "/embeddingLayerNorm/scale"),
                                                   axis: -1,
                                                   epsilon: 0.001)
        self.embeddingDropout = Dropout<Float>(probability: Double(config.dropoutProbability))
        self.transformer = TransformerEncoder(reader: reader, config: config, scope: scope + "/transformer")
    }
}
```

@BradLarson One thing to keep in mind while working on checkpointing is that we also need optimizer saving and loading. Stateful optimizers like Adam and RAdam will need to have their state serialized and deserialized to support continued training on a saved model. Initializing a stateful optimizer from scratch will grossly overshoot the gradient calculations, producing a spike in the train/val loss. I believe there are fine-tuning situations where this is also desirable. It can be preferable to start the optimizer's moments from the last pre-training epoch rather than from 0 to prevent the aforementioned gradient overshooting. I am currently working on manually serializing and deserializing my optimizer's properties so that I can continue training on a model.

**Comment by BradLarson** [comment]
I've captured the need for optimizer, training loop, etc. state checkpointing in [swift-models issue #726](https://github.com/tensorflow/swift-models/issues/726), but the presence of our checkpointing APIs and the pending completion of [a higher-level interface for model checkpoints](https://github.com/tensorflow/swift-models/pull/631) would seem to address the core needs of this issue. As a result, I'm going to mark this as completed here in favor of further discussion about checkpointing over at [tensorflow/swift-models](https://github.com/tensorflow/swift-models).
